src/core.rs

```diff
@@ -8,15 +8,60 @@
 
 use std::fmt;
 
+struct TreeWriter<'a, Item: ?Sized + fmt::Debug + 'a> {
+    items: Vec<&'a Predicate<Item>>,
+}
+
 /// Trait for generically evaluating a type against a dynamically created
 /// predicate function.
 ///
 /// The exact meaning of `eval` depends on the situation, but will usually
 /// mean that the evaluated item is in some sort of pre-defined set.  This is
 /// different from `Ord` and `Eq` in that an `item` will almost never be the
 /// same type as the implementing `Predicate` type.
-pub trait Predicate<Item: ?Sized>: fmt::Display {
+pub trait Predicate<Item: ?Sized + fmt::Debug>: fmt::Display {
     /// Execute this `Predicate` against `variable`, returning the resulting
     /// boolean.
     fn eval(&self, variable: &Item) -> bool;
+
+    /// TODO
+    fn flatten<'a, 'b>(&'a self, _vec: &'b mut Vec<&'a Predicate<Item>>) {
+        unimplemented!()
+    }
+
+    /// TODO
+    fn stringify(&self, _variable: &Item) -> String {
+        unimplemented!()
+    }
+
+    /// TODO
+    #[cfg(feature = "term-table")]
+    fn tree_eval(&self, variable: &Item) -> String {
```

Comment:
We need a way to programmatically know the result of the evaluation, so a caller like assert_cmd can decide whether to panic or not.